### PR TITLE
[MRG] forkless mode for debugging

### LINKED ISSUE #1802: Fork-free mode for debugging
So that we can use `ipdb` to debug Toil

### PR DESCRIPTION
**Pull request by mr-c**
Closes #1802

### CODE REVIEW

**Comment by mr-c** [comment]
Example usage

```
$ ipdb venv/bin/toil-cwl-runner --forkless --logDebug --disableCaching src/toil/test/cwl/revsort.cwl src/toil/test/cwl/revsort-job.json
ipdb> break toil/cwl/cwltoil.py:427
ipdb> continue
Breakpoint 1 at src/toil/cwl/cwltoil.py:427
ipdb> self.cwljob
{u'input': (u'input', {'input': ordereddict([('class', 'File'), ('location', 'toilfs:F/w/tmpArgqF8.tmp'), ('size', 1111), ('basename', 'whale.txt'), ('nameroot', 'whale'), ('nameext', '.txt')]), u'reverse_sort': True})}
```

**Comment by brainstorm** [comment]
As always, CWL version pinning bites, grrr, fixing locally...:

```
(py2) brainstorm:toil romanvg$ ipdb toil-cwl-runner --forkless --logDebug --disableCaching src/toil/test/cwl/revsort.cwl src/toil/test/cwl/revsort-job.json
Error: toil-cwl-runner does not exist
(py2) brainstorm:toil romanvg$ toil-cwl-runner
Traceback (most recent call last):
  File "/Users/romanvg/.miniconda/envs/py2/bin/toil-cwl-runner", line 11, in <module>
    load_entry_point('toil', 'console_scripts', 'toil-cwl-runner')()
  File "/Users/romanvg/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 563, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/Users/romanvg/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2651, in load_entry_point
    return ep.load()
  File "/Users/romanvg/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2304, in load
    self.require(*args, **kwargs)
  File "/Users/romanvg/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2327, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/Users/romanvg/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 862, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (cwltool 1.0.20170811195303 (/Users/romanvg/.miniconda/envs/py2/lib/python2.7/site-packages), Requirement.parse('cwltool==1.0.20170822192924'))
```

**Comment by brainstorm** [comment]
Ohh, man, this is beautiful:

```
ipdb> workdir
'/var/folders/wz/__dd9trs0kl4jb3rs83704y80000gn/T/tmpDIanEf'
```

Yeaaaahhh :D

**Comment by mr-c** [comment]
@brainstorm Glad it worked for you, enjoy!

File: src/toil/batchSystems/singleMachine.py [hunk 1]
```diff
@@ -151,28 +157,52 @@ def worker(self, inputQueue):
                                   jobCores)
                         with self.coreFractions.acquisitionOf(coreFractions):
                             with self.disk.acquisitionOf(jobDisk):
-                                startTime = time.time() #Time job is started
-                                with self.popenLock:
-                                    popen = subprocess.Popen(jobCommand,
+                                startTime = time.time()  # Time job is started
```
Comment (diekhans) [inline src/toil/batchSystems/singleMachine.py]: Could this be split into multiple functions? Function is way too long and nested

**Comment by diekhans** [comment]
I don't see any use documentation at all.  Also forkless is a very confusing name, as the actually job
is still exec-ed

**Comment by diekhans** [comment]
I would favor calling this --debug-worker , because that is what this is enabling.

There is a larger problem with debugging the actually application

**Comment by mr-c** [comment]
@diekhans I've updated the option name to `--debug-worker`, thank you for the suggestion.

**Comment by ejacox** [comment]
Looks like a transient error. I want to refactor the code a bit and I will retest it.

**Comment by ejacox** [comment]
@mr-c, we can merge this if you are OK with the changes I made.

File: src/toil/batchSystems/singleMachine.py [hunk 2]
```diff
@@ -122,19 +123,62 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
...
+    def _runWorker(self, jobCommand, jobID, environment):
+        """
+        Run the jobCommand using the worker and wait for it to finish.
+        The worker is forked unless it is a '_toil_worker' job and
+        debugWorker is True.
+        """
+        startTime = time.time()  # Time job is started
+        popen = None
+        statusCode = None
+        forkWorker = not self.debugWorker or "_toil_worker" not in jobCommand
```
Comment (mr-c) [inline src/toil/batchSystems/singleMachine.py]: We always want to fork (default behavior) unless debugworker && _toil_worker in jobcommand

```
forkWorker = not (self.debugWorker and "_toil_worker" not in jobCommand)
```

I agree that not forking when running locally could be a useful optimization, but that should be tested separately.

File: src/toil/batchSystems/singleMachine.py [hunk 3]
```diff
@@ -143,7 +143,7 @@ def _runWorker(self, jobCommand, jobID, environment):
         startTime = time.time()  # Time job is started
         popen = None
         statusCode = None
-        forkWorker = not self.debugWorker or "_toil_worker" not in jobCommand
+        forkWorker = not (self.debugWorker and "_toil_worker" not in jobCommand)
         if forkWorker:
```
Comment (ejacox) [inline src/toil/batchSystems/singleMachine.py]: I thought it was threading only if "_toil_work" is in jobCommand. Did you want to fork for any non '_toil_worker' command, no matter the setting for self.debugWorker? If so, the second 'not' should be removed. Either way, a comment explaining the logic might be useful.

Comment (mr-c) [reply src/toil/batchSystems/singleMachine.py]: The original behaviour was to always call `popen`, my hack should only be applied if and only if `_toil_worker` is present **and** `--debugWorker` is specified.

Did you see https://github.com/BD2KGenomics/toil/pull/1910#discussion_r142006836 ?

I'll add a comment with the above.

Comment (ejacox) [reply src/toil/batchSystems/singleMachine.py]: The second 'not' needs to removed. Do you agree @mr-c? In parentheses, is the condition for threading, which is if self.debugWorker is True and "_toil_worker" is in jobCommand.

**Comment by ejacox** [comment]
Thank you @mr-c and @brainstorm.

**Pull request merged**